### DISCUSSION TITLE
CloudWatch Log Group Retention Period

### DISCUSSION THREAD

**Discussion by PraxisTwirly** [cloud-custodian/cloud-custodian#7805] [AWS] [answered]
I wrote a policy to identify CloudWatch log groups that have the retention period set to Never expire, but it is returning unexpected results.  I expect the following policy to return all log groups that have retention set to Never expire:

```
policies:
  - name: cloudwatch-log-groups-never-expire
    resource: aws.log-group
    filters:
      - type: value
        key: retentionInDays
        value: absent
```

When I do this, I get no matching records when I expected 1 matching record.  I verified the AWS CLI does not return a retentionInDays field when Never expire is set.  When I reviewed the resources.json output from my Cloud Custodian run, the resource shows retentionInDays: 90.

I verified if I set any other retention period, the retentionInDays matches as expected between Cloud Custodian and AWS CLI.  It is only when I set Never expire that Cloud Custodian returns 90 instead of no record at all.

Here is an example of the output from AWS CLI when Never expire is set:

```
{
            "logGroupName": "/aws-glue/crawlers",
            "creationTime": 1559847036116,
            "metricFilterCount": 0,
            "arn": "arn:aws:logs:xx-region-x:xxxxxxxxxxxxxxx:log-group:/aws-glue/crawlers:*",
            "storedBytes": 0
}
```        

Here is an example of the resources.json file output from Cloud Custodian that incorrectly shows 90 days if I modify the policy to return records where retentionInDays is present instead of absent:

```
{
    "logGroupName": "/aws-glue/crawlers",
    "creationTime": 1559847036116,
    "retentionInDays": 90,
    "metricFilterCount": 0,
    "arn": "arn:aws:logs:xx-region-x:xxxxxxxxxxxxxxx:log-group:/aws-glue/crawlers:*",
    "storedBytes": 0,
    "Tags": [],
    "c7n:MatchedFilters": [
      "retentionInDays"
    ]
  }
```

Here are screenshots of the AWS console, CLI command output, and Cloud Custodian resources.json output.

![image](https://user-images.githubusercontent.com/114627161/192910038-84471554-6c28-4400-8601-0cfdde5b57f1.png)

Why does this matter?  CloudWatch log groups that never expire can result in excessive costs and I would like to enforce that all log groups have at least a maximum default retention period of some arbitrary value.  I am able to do this with AWS Config and Lambda functions successfully, but I am unable to do this with Cloud Custodian.

Please note I'm not sure if this is an error in my policy or a bug, so I started the discussion to get feedback from the community before filing an issue.

**Comment by ajkerrigan** [comment, marked as answer]
Custodian will cache resource data for 15 minutes by default, to avoid hammering cloud provider APIs when it needs the same data multiple times over a short period. So if you were testing policy runs and tweaking retention at the same time, it's likely that Custodian was working with stale data.

To take that behavior out of the picture, you can add `--cache-period 0` to your `custodian run` command. Under the hood, Custodian is making the same call that your `describe-log-groups` CLI command does. So setting aside any timing/caching issues, the retention values should match across the board.

**Reply by PraxisTwirly** [reply]
I set the --cache-period to 0 and the command returned the expected results.  Thank you for your help!
